src/libraries/Microsoft.Extensions.Logging.Abstractions/src/FormattedLogValues.cs

```diff
@@ -103,7 +106,14 @@ public override string ToString()
                 return _originalMessage;
             }
 
-            return _formatter.Format(_values);
+            if (_cachedToString is null)
+            {
+                // caching the formatted string to avoid repeated formatting and allocations.
+                // FormattedLogValues is readonly struct, use Unsafe.AsRef to mutate the cached string.
+                Unsafe.AsRef(in _cachedToString) = _formatter.Format(_values);
```

Comment:
Interesting. I take it that we regard this to be safe because the result of `_formatter.Format` is idempotent? In STJ we ended up going the opposite path in a [similar scenario](https://github.com/dotnet/runtime/pull/114071) because we wanted the struct to remain provably immutable. I'd be curious if others could weigh in on this technique and whether it's safe to use elsewhere. cc @jkotas @stephentoub

Reply:
I removed the readonly on FormattedLogValues